test/micro/org/openjdk/bench/jdk/incubator/foreign/LoopOverNonConstant.java

```diff
@@ -116,6 +132,16 @@ public int unsafe_loop() {
         return res;
     }
 
+    @Benchmark
+    public int global_segment_loop() {
+        // In fact, we operate inside `segment`
+        int res = 0;
+        for (int i = 0; i < ELEM_SIZE; i ++) {
+            res += (int) VH_int.get(globalRestrictedSegment, segment_addr_idx + i);
```

Comment:
This looks wrong. you are passing an absolute address to a "logical index" argument.  I see that you are attempting to divide the segment address by the carrier size, and that kind offset things, but still leaves you with suboptimal performances.

Reply:
If you want to access memory this way, you need a different VarHandle:

```
VarHandle absolute_int_VH = MemoryHandles.varHandle(int.class, ByteOrder.nativeOrder());
```
Or use one of the static accessors:

```
MemoryAccess.getIntAtOffset(globalRestrictedSegment, segment_addr_idx + i);
```